File: abc054c/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

var sc = bufio.NewScanner(os.Stdin)

const nmax = 8

var graph = [nmax][nmax]bool{}

func nextInt() int {
	sc.Scan()
	i, e := strconv.Atoi(sc.Text())
	if e != nil {
		panic(e)
	}
	return i
}

func main() {
	sc.Split(bufio.ScanWords)

	n := nextInt()
	m := nextInt()

	for i := 0; i < m; i++ {
		a := nextInt()
		b := nextInt()
		graph[a-1][b-1] = true
		graph[b-1][a-1] = true
	}

	var visited = [nmax]bool{}
	visited[0] = true

	ans := dfs(0, n, visited[:])
	fmt.Println(ans)
}

func dfs(v, N int, visited []bool) int {
	allVisited := true
	for i := 0; i < N; i++ {
		if !visited[i] {
			allVisited = false
		}
	}

	if allVisited {
		return 1
	}

	ret := 0

	for i := 0; i < N; i++ {
		if !graph[v][i] {
			// 頂点がつながってないので無視
			continue
		}
		if visited[i] {
			// すでに訪れたので無視（今回の問題は一筆書き）
			continue
		}

		// 上記のチェックに引っかからなかった
		// 到達可能なためtrue
		visited[i] = true
		// さらに深いポイントを探索していく
		ret += dfs(i, N, visited)
		// このルートは探索が終わったのでfalseに戻す
		visited[i] = false
	}

	return ret
}

File: abc054c/rust/src/main.rs
const NMAX : usize = 8;

fn main() {
    let vec = read_line();
    let n = vec[0];
    let m = vec[1];

    let mut a = vec![];
    let mut b = vec![];
    for _ in 0..m {
        let vec = read_line();
        a.push(vec[0] - 1);
        b.push(vec[1] - 1);
    }

    let mut graph = vec![vec![false; NMAX]; NMAX];

    // 移動できるポイントにフラグを立てる
    for i in 0..m as usize {
        graph[a[i] as usize][b[i] as usize] = true;
        graph[b[i] as usize][a[i] as usize] = true;
    }

    let mut visited = vec![false; NMAX];
    visited[0] = true;

    let ans = dfs(0, n as usize, &mut visited, &mut graph);
    println!("{}", ans);
}

fn dfs(v : usize, n : usize, visited : &mut Vec<bool>, graph : &mut Vec<Vec<bool>>) -> i32 {
    let mut all_visited = true;
    for i in &*visited { // まじで！？
        if *i {
            all_visited = false;
        }
    }

    if all_visited {
        return 1;
    }

    let mut ret = 0;
    for i in 0..n as usize {
        if !graph[v][i] {
            continue;
        } else if visited[i] {
            continue;
        }

        visited[i] = true;
        ret += dfs(i, n, visited, graph);
        visited[i] = false;
    }

    return ret;
}

fn read_line() -> Vec<i32> {
    let mut s = String::new();
    std::io::stdin().read_line(&mut s).ok();
    let v = s.trim()
        .split_whitespace()
        .map(|e| e.parse().ok().unwrap())
        .collect();
    return v;
}
